backend/app/libs/tasks/get_assets.py

```diff
@@ -354,36 +364,82 @@ def setup_init_tasks(**_):
     reload_whitelist.apply_async()
 
 
-@celery_app.task
-def reload_treasuries_data():
+LOCK_EXPIRE = 60 * 30  # lock expires in 20 minutes
+
+
+@contextmanager
+def redis_lock(lock_id: str, oid: Any):
+    status = db.setnx(lock_id, oid)
+    if status:
+        db.expire(lock_id, LOCK_EXPIRE)
+    yield status
+
+
+def check_lock(lock_id: str, task_id: str, oid, logger):
+    with redis_lock(lock_id, oid) as acquired:
+        if not acquired:
+            logger.info("task %s already running, revoking...", task_id)
+            celery_app.control.revoke(task_id, terminate=True)
+
+
+def set_new_task(task_id: str):
+    db.set("curr_task", task_id)
+    if db.ttl("curr_task") <= 0:
+        db.expire("curr_task", LOCK_EXPIRE)
+
+
+@celery_app.task(bind=True)
+def reload_treasuries_stats(self):
+    logger = get_task_logger(self.request.id)
+    check_lock(f"{self.name}-lock", self.request.id, celery_app.oid, logger)
+    set_new_task(self.request.id)
+
     end_date: datetime = today(UTC)
     start_date: datetime = end_date - timedelta(days=365)
 
     start = start_date.isoformat()[:10]
     end = end_date.isoformat()[:10]
 
-    for treasury_metadata in retrieve_treasuries_metadata():
+    treasuries = retrieve_treasuries_metadata(db)
```

Comment:
I'd prefer that we add tests for `store_treasuries_metadata` and `retrieve_treasuries_metadata` in this PR because these are needed to be confident merging in main.

Let's test every aspect of a given PR as a unit of change. Refactoring can still be done but that's an independent concern.